### ISSUE TITLE
Support FastSurferCNN on Apple M1

### ISSUE THREAD

**Issue by emilljungberg** [Deep-MI/FastSurfer#109]
I'm trying to run FastSurfer CNN for segmentation on my MacBook Pro with the Apple M1 chip. I was able to build to docker container for CPU using your instructions (example 4) by adding in `--platform linux/x86_64`, but there seems to be some problem running the program. When I execute the following (note the added platform here as well)
```
docker run --platform linux/x86_64 -v $SUBJECT_DIR:/data \
           -v $outpath:/output \
           --rm --user 501 fastsurfercnn:cpu \
           --i_dir /data \
           --in_name mri/mprage.mgz \
           --o_dir /output \
           --out_name mri/aparc.DKTatlas+aseg.deep.mgz \
           --log deep_surfer.log \
           --no_cuda
```
I get the following output
```
Total Dataset Size is 1
Running Fast Surfer on sub01
Reading volume /data/sub01/mri/mprage.mgz
Conforming image to UCHAR, RAS orientation, and 1mm isotropic voxels
Input:    min: 0  max: 1851
rescale:  min: 0.0  max: 831.099  scale: 0.30682265289694727
Output:   min: 0.0  max: 255.0
Cuda available: False, # Available GPUS: 0, Cuda user disabled (--no_cuda flag): True, --> Using device: cpu
```
The program creates the correct output folder structure but there is no output files in there. Any clue what might be going wrong?

## Environment
 - FastSurfer Version: Pulled from Github 2022/02/25
 - FreeSurfer Version: 7.2.0 (know this is higher than recommended but not used for the segmentation right?)
 - OS: MacOS 12.1
 - CPU: Apple M1 Pro

**Comment by m-reuter** [comment]
Hi Emil, 
we do not support the M chips yet. Probably in the future. Maybe it is easy or maybe not. We can keep this issue open, for people to post if they have success getting it running. 

And also 7.2 is not only not recommended, it is not supported. Use 6.0 for now. But you are right, for only the segmentation it is not used and should not matter.

**Comment by LeHenschel** [comment]
Maybe also a memory issue (see issue https://github.com/Deep-MI/FastSurfer/issues/40)? In order to run FastSurfer you need around 10 GB RAM, but Docker Desktop on MAC is set to use 2 GB runtime memory by default. You can overwrite the settings under Docker Desktop --> Preferences --> Resources --> Advanced, slide the bar under Memory to 10 GB; see: https://docs.docker.com/docker-for-mac/ for details.

Kind regards,
Leonie

**Comment by emilljungberg** [comment]
It seems to work with increasing the RAM, set it to 20 GB to try. It runs now and seems to use about 7.2 GB. Let's see how long it takes to finish.

**Comment by emilljungberg** [comment]
Update: I actually managed to run FastSurferCNN natively on my Mac without using docker. There was a massive speed up from not using Docker. Results looks good! Thanks!

**Comment by m-reuter** [comment]
Thanks for the update. Can you let readers know how you did it? Anything special you had to do or did it run out of the box? Thanks

**Comment by emilljungberg** [comment]
It worked straight out of the box actually. Just installed the pre-reqs as instructed and it worked!

**Comment by m-reuter** [comment]
**FastSurferCNN Segmentation on CPU only:**
I got a docker with aarch64 running on Apple M1 by pointing the docker file to the correct miniconda:
`ARG CONDA_FILE=Miniconda3-py38_4.11.0-Linux-aarch64.sh`
I also had to modify some dependencies versions in the fastsurfer_env_cpu.yml and drop the pip stuff (which is needed for recon_surf and not for the CNN). Increasing Docker memory to 15GB I was able to segment an image in below 30 mins (using the dev version). 

Do not try to build an AMD64 docker, it will work but will be extremely slow (1 batch takes hours). 

In order to get recon_surf running on ARM64 (=aarch64) we would need to install the pip dependencies also and compile FreeSurfer for ARM, for that see: https://github.com/freesurfer/freesurfer/issues/964

**Comment by m-reuter** [comment]
I now managed a native install on a MacBook Pro M1-max mainly using  pip3, changed a few version numbers slightly in requirements.txt and had to use brew to install a newer bash. I also solved issues with installing h5py by: 
```
brew install hdf5
export HDF5_DIR="$(brew --prefix hdf5)"
pip3 install --no-binary=h5py h5py
```

Run time for one full segmentation on CPU (multi threads) with current dev branch is only 5 mins and 25 seconds now, not bad. This is all three views and view aggregation. So don't use docker on M1, not sure why docker is so much slower, it was set to use 5 cores, using 8 made it even slower.

**Comment by m-reuter** [comment]
In order to get this running on the Apple GPU with MPS support in PyTorch nightly we would need `aten::max_unpool2d` natively on MPS (otherwise it will revert to CPU). I put in a request for accelerated implementation here:
https://github.com/pytorch/pytorch/issues/77764#issuecomment-1215920185
Please consider upvoting that there. 

Currently eval.py with MPS is not yet working.

**Comment by m-reuter** [comment]
I traced the MPS issues (code producing -inf values) a little further. The first big difference between the CPU and the MPS code occurs in nn.BatchNorm2d. See my Issue here: 
https://github.com/pytorch/pytorch/issues/84962

**Comment by m-reuter** [comment]
The BatchNorm2 differences are solved in MacOS Ventura. Another issue https://github.com/pytorch/pytorch/issues/85230 is that loading the weights directly to mps device fails. There is a workaround to move the model to cpu, load checkpoint and then move the model back to mps. And ...

Tadaaa :-) mps works! It takes 2 min 40 seconds for all three networks (VINN) on MPS and view aggregation on cpu. Nice. I will put in a pull request soon, so that this can go into the next release 2.0. And don't forget to up-vote support for max_unpool2d here: https://github.com/pytorch/pytorch/issues/77764#issuecomment-1215920185 . should get even faster once that is native on MPS.

**Comment by m-reuter** [comment]
Has been solved with #191
